modules/rostests/apitests/ntdll/NtStartProfile.c

```diff
@@ -0,0 +1,143 @@
+void
+LoopFunction(
+    _Inout_updates_all_(BufferSize) volatile ULONG *Buffer,
+    _In_ ULONG BufferSizeInElements,
+    _In_ ULONG LoopCount)
+{
+    ULONG i;
+    ULONG Offset;
+
+    for (i = 0; i < LoopCount; i++)
+    {
+        for (Offset = 0; Offset < BufferSizeInElements; Offset++)
+        {
+            Buffer[Offset]++;
+        }
+    }
+}
+
+static
+void
+ProfileLoopFunction(
+    _In_ LOOP_FUNCTION *Function,
+    _Out_writes_bytes_(BufferSize) PULONG Buffer,
+    _In_ ULONG BufferSize,
+    _In_range_(0, BufferSize / sizeof(ULONG)) ULONG BufferOffset
+    )
+{
+    NTSTATUS Status;
+    HANDLE ProfileHandle;
+    ULONG Buffer1Value;
+
+    Status = NtCreateProfile(&ProfileHandle,
+                             NtCurrentProcess(),
+                             (PVOID)((ULONG_PTR)Function - LOOP_FUNCTION_SIZE),
+                             3 * LOOP_FUNCTION_SIZE,
+                             LOOP_FUNCTION_SIZE_SHIFT,
+                             Buffer,
+                             BufferSize,
+                             ProfileTime,
+                             SystemAffinityMask);
+    ok_hex(Status, STATUS_SUCCESS);
+    if (!NT_SUCCESS(Status))
+    {
+        skip("Failed to create profile\n");
+        return;
+    }
+
+    Status = NtStartProfile(ProfileHandle);
+    ok_hex(Status, STATUS_SUCCESS);
+
+    /* Can't validate Buffer contents here, since we don't know what's next to our function */
+
+    /*
+     * This takes around 10-12 seconds on my machine, which is not ideal.
+     * But on my Win2003 VM it only results in counts of 10-12,
+     * which means we can't really make it shorter.
+     */
+
+    /* Run a long loop */
+    Function(DummyBuffer,
+             RTL_NUMBER_OF(DummyBuffer),
+             1000000);
+
+    /* The buffer should get live updates */
+    Buffer1Value = Buffer[BufferOffset];
+    ok(Buffer1Value != 0, "Buffer[%lu] = %lu\n", BufferOffset, Buffer1Value);
+
+    /* Run a shorter loop, we should see a smaller increase */
+    Function(DummyBuffer,
+             RTL_NUMBER_OF(DummyBuffer),
+             200000);
+
+    ok(Buffer[BufferOffset] > Buffer1Value,
+       "Buffer[%lu] = %lu, expected more than %lu\n",
+       BufferOffset, Buffer[BufferOffset], Buffer1Value);
+
+    Status = NtStopProfile(ProfileHandle);
+    ok_hex(Status, STATUS_SUCCESS);
+
+    /* The expectation is that Buffer[BufferOffset] is somewhere around 20% larger than Buffer1Value.
+     * Allow anywhere from one more count to twice as many to make the test robust.
+     */
+    ok(Buffer[BufferOffset] > Buffer1Value,
+       "Buffer[%lu] = %lu, expected more than %lu\n", BufferOffset, Buffer[BufferOffset], Buffer1Value);
+    ok(Buffer[BufferOffset] < 2 * Buffer1Value,
+       "Buffer[%lu] = %lu, expected less than %lu\n", BufferOffset, Buffer[BufferOffset], 2 * Buffer1Value);
+
+    trace("Buffer1Value = %lu\n", Buffer1Value);
+    trace("Buffer[%lu] = %lu\n", BufferOffset - 1, Buffer[BufferOffset - 1]);
```

Comment:
Since it's a buffer with three elements (not obvious from the function signature now), I was just printing all the values to make sure things were behaving as expected.
I guess I might do a loop to print all non-zero values and their corresponding addresses, that's probably more useful for future debugging.